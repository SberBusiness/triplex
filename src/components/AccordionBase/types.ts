import React from 'react';
import {TAriaHTMLAttributes} from '@sberbusiness/triplex/utils/HTML/AriaAttributes';

/**
 * Свойства, передаваемые в рендер-функцию body.
 */
export interface IAccordionBodyProvideProps {
    // Анимируется в текущий момент.
    animating: boolean;
    // Флаг раскрытости.
    isOpen: boolean;
    // Функция открыть/закрыть.
    toggle: () => void;
}

/**
 * Свойства, передаваемые в рендер-функцию header.
 */
export interface IAccordionHeaderProvideProps {
    // Анимируется в текущий момент.
    animating: boolean;
    // Aria-атрибуты a11y.
    ariaAttributes: TAriaHTMLAttributes;
    // флаг раскрытости.
    isOpen: boolean;
    // Функция открыть/закрыть.
    toggle: () => void;
}

/**
 * Свойства AccordionBase.
 */
export interface IAccordionBaseProps {
    children?: never;
    // ClassName для ExpandAnimation.
    expandAnimationClassName?: string;
    // Флаг раскрытости.
    isOpen?: boolean;
    // Обработчик открытия/закрытия.
    onToggle?: (isOpen: boolean) => void;
    // Рендер-функция body.
    renderBody: (props: IAccordionBodyProvideProps) => React.ReactNode;
    // Рендер-функция header.
    renderHeader: (props: IAccordionHeaderProvideProps) => JSX.Element;
    // Функция открыть/закрыть.
    toggle?: (isOpen: boolean) => void;
}
