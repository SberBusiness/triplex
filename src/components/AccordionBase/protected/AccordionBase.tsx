import React from 'react';
import {ExpandAnimation} from '@sberbusiness/triplex/components/ExpandAnimation/ExpandAnimation';
import {uniqueId} from '@sberbusiness/triplex/utils/uniqueId';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {IAccordionBaseProps} from '../types';

/**
 * Состояние AccordionBase.
 *
 * @prop {boolean} animating Анимируется в текущий момент.
 * @prop {boolean} controlled Управляемый аккордеон или неконтролируемый.
 * @prop {boolean} isOpen Флаг раскрытости.
 */
interface IAccordionBaseState {
    animating: boolean;
    controlled: boolean;
    isOpen: boolean;
}

/**
 * Базовый компонент Accordion.
 */
export class AccordionBase extends React.Component<IAccordionBaseProps, IAccordionBaseState> {
    public displayName = 'AccordionBase';

    // Уникальный id, для передачи a11y aria-атрибутов в header и body.
    private instanceId = `AccordionBase-${uniqueId()}`;

    constructor(props: IAccordionBaseProps) {
        super(props);

        this.state = {
            animating: false,
            controlled: typeof props.isOpen !== 'undefined',
            isOpen: Boolean(props.isOpen),
        };
    }

    public componentDidUpdate(prevProps: IAccordionBaseProps): void {
        const {isOpen} = this.props;
        const {isOpen: prevIsOpen} = prevProps;

        if (isOpen !== prevIsOpen) {
            this.setState({animating: true});
        }
    }

    public render(): React.ReactElement {
        const header = this.renderHeader();
        const body = this.renderBody();

        return (
            <>
                {header}
                {body}
            </>
        );
    }

    private renderHeader = () => {
        const {renderHeader, isOpen: isOpenProps} = this.props;
        const {animating, controlled, isOpen: isOpenState} = this.state;
        const isOpen = Boolean(controlled ? isOpenProps : isOpenState);

        return renderHeader({
            animating,
            ariaAttributes: {
                controls: this.instanceId,
                expanded: isOpen.toString(),
            },
            isOpen,
            toggle: this.toggle,
        });
    };

    /**
     * Раскрывает/скрывает контент аккордеона.
     */
    private toggle = () => {
        const {controlled, isOpen: isOpenState} = this.state;
        const {toggle, isOpen: isOpenProps} = this.props;
        const isOpen = Boolean(controlled ? isOpenProps : isOpenState);

        if (controlled) {
            toggle && toggle(!isOpen);
        } else {
            this.setState((prevState) => ({
                animating: true,
                isOpen: !prevState.isOpen,
            }));
        }
    };

    private handleAnimationEnd = () => {
        const {controlled, isOpen: isOpenState} = this.state;
        const {onToggle, isOpen: isOpenProps} = this.props;
        const isOpen = Boolean(controlled ? isOpenProps : isOpenState);

        this.setState({animating: false}, () => {
            if (onToggle) {
                onToggle(isOpen);
            }
        });
    };

    private renderBody = () => {
        const {expandAnimationClassName, renderBody, isOpen: isOpenProps} = this.props;
        const {animating, controlled, isOpen: isOpenState} = this.state;
        const isOpen = Boolean(controlled ? isOpenProps : isOpenState);

        return (
            <ExpandAnimation
                className={classnames(expandAnimationClassName, {
                    'cssClass[globalExpandAnimationAnimating]': animating,
                    'cssClass[globalExpandAnimationOpened]': isOpen,
                })}
                expanded={isOpen}
                onEnd={this.handleAnimationEnd}
                id={this.instanceId}
            >
                {renderBody({
                    animating,
                    isOpen,
                    toggle: this.toggle,
                })}
            </ExpandAnimation>
        );
    };
}
