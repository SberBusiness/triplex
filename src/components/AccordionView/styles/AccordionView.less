@import url('src/styles/style');

.accordionView {
    background: var(--triplex-AccordionView-Background);
    border-radius: @radius-16;

    .accordionViewHeaderWrapper {
        display: flex;
        justify-content: space-between;
        outline: none;
        border: 1px solid var(--triplex-AccordionView-BorderColor_Default);
        border-radius: @radius-16;
        padding: 19px 31px;
        cursor: pointer;

        .accordionViewHeader {
            .h5-header-form(@color-graphite-01);

            color: var(--triplex-AccordionView-Header_Color);
            flex-grow: 1;
            overflow: hidden;
            white-space: nowrap;
            text-overflow: ellipsis;
        }

        .caretIcon {
            transition: transform 0.3s ease-in-out;
            margin: 4px 0 4px 20px;
            flex-shrink: 0;
        }

        &:not(.focusedByClick):focus {
            border-color: var(--triplex-AccordionView-BorderColor_Focus);
        }

        &.animating,
        &.opened {
            border-radius: @radius-16 @radius-16 0 0;

            .caretIcon {
                transform: rotate(-180deg);
            }
        }
    }

    .accordionViewBody {
        padding: 24px 31px 23px;
    }

    .accordionViewContent {
        box-sizing: content-box;
        visibility: hidden;

        &.globalExpandAnimationAnimating,
        &.globalExpandAnimationOpened {
            border: 1px solid var(--triplex-AccordionView-BorderColor_Default);
            border-top: none;
            border-radius: 0 0 @metric-16 @metric-16;
            visibility: visible;
        }
    }
}

@media (max-width: @screen-sm-max) {
    .accordionView {
        .accordionViewHeaderWrapper,
        .accordionViewBody {
            padding: 15px;
        }
    }
}