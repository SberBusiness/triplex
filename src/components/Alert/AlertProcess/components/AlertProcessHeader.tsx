import React from 'react';

/** Свойства компонента AlertProcessHeader. */
export interface IAlertProcessHeaderProps extends React.HTMLAttributes<HTMLDivElement> {}

/** Заголовок компонента AlertProcess. */
export const AlertProcessHeader: React.FC<IAlertProcessHeaderProps> = ({children, ...htmlAttributes}) => (
    <div {...htmlAttributes} className="cssClass[alertProcessContentHeader]">
        {children}
    </div>
);

AlertProcessHeader.displayName = 'AlertProcessHeader';
