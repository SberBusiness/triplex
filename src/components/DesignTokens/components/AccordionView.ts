import {TDesignTokenValue, TDesignTokenValues} from '@sberbusiness/triplex/components/DesignTokens/types/DesignTokenTypes';

// Название токенов компонента AccordionView.
export const designTokensComponentsAccordionViewKeys = ['BorderColor_Default', 'BorderColor_Focus', 'Header_Color'] as const;
// Тип, содержащий названия токенов компонента AccordionView.
export type TDesignTokensComponentsAccordionViewKeys = (typeof designTokensComponentsAccordionViewKeys)[number];
// Тип, содержащий названия токенов компонента AccordionView и их значения.
export type TDesignTokensComponentsAccordionViewValue = Record<TDesignTokensComponentsAccordionViewKeys, TDesignTokenValue>;
// Тип, содержащий названия токенов компонента AccordionView и их значения в светлой и темной теме.
export type TDesignTokensComponentsAccordionViewValues = Record<TDesignTokensComponentsAccordionViewKeys, TDesignTokenValues>;
// Тип локальных токенов компонента AccordionView.
export type TDesignTokensComponentsAccordionView = {AccordionView: TDesignTokensComponentsAccordionViewValue};

// Токены компонента AccordionView в светлой и темной темах.
export const AccordionView_Tokens: TDesignTokensComponentsAccordionViewValues = {
    BorderColor_Default: [{ref: 'Neutral.500'}, {ref: 'Basic.500'}], // var(--triplex-AccordionView-BorderColor_Default)
    BorderColor_Focus: [{ref: 'Accent.500'}, {ref: 'Accent.500'}], // var(--triplex-AccordionView-BorderColor_Focus)
    Header_Color: [{ref: 'Basic.700'}, {ref: 'Neutral.100'}], // var(--triplex-AccordionView-Header_Color)
};
