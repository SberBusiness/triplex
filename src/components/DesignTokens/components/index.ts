export * from './AccordionForm';
export * from './AccordionView';
export * from './AlertContext';
export * from './AlertProcess';
export * from './Body';
export * from './Button';
export * from './Calendar';
export * from './Card';
export * from './Checkbox';
export * from './Chip';
export * from './Confirm';
export * from './Divider';
export * from './Dropdown';
export * from './DropdownList';
export * from './DropdownMobile';
export * from './DropdownMobileList';
export * from './Footer';
export * from './FormField';
export * from './Header';
export * from './Input';
export * from './Label';
export * from './LightBox';
export * from './Link';
export * from './ListItem';
export * from './ListItemControlsButton';
export * from './ListMaster';
export * from './Marker';
export * from './MarkerStatus';
export * from './Multiselect';
export * from './Notification';
export * from './Pagination';
export * from './Radio';
export * from './SegmentedControl';
export * from './Select';
export * from './Skeleton';
export * from './Slider';
export * from './SmallInput';
export * from './SMSInput';
export * from './Spinner';
export * from './Spoiler';
export * from './Step';
export * from './Stepper';
export * from './TableBasic';
export * from './TableBasicSettings';
export * from './Tabs';
export * from './TabsFolder';
export * from './TabsLine';
export * from './Tag';
export * from './Text';
export * from './Title';
export * from './Tooltip';
export * from './TopOverlay';
export * from './UploadZone';
export * from './Widget';
