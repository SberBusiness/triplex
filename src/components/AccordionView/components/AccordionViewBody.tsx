import React from 'react';
import {IAccordionBodyProvideProps} from '@sberbusiness/triplex/components/AccordionBase/types';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';

/** Свойства компонента AccordionViewBody. */
export interface IAccordionViewBodyProps extends Partial<IAccordionBodyProvideProps>, React.HTMLAttributes<HTMLDivElement> {}

/** Тело компонента аккордеон. */
export class AccordionViewBody extends React.PureComponent<IAccordionViewBodyProps> {
    public static displayName = 'AccordionViewBody';

    public render(): JSX.Element {
        const {children, className} = this.props;
        return (
            <div className={classnames(className, 'cssClass[accordionViewBody]')} {...this.getHtmlDivAttributes()}>
                {children}
            </div>
        );
    }

    private getHtmlDivAttributes = (): React.HTMLAttributes<HTMLDivElement> => {
        const {animating, className, isOpen, toggle, ...htmlDivAttributes} = this.props;
        return htmlDivAttributes;
    };
}
