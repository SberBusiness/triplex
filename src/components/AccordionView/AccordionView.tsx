import React from 'react';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {
    AccordionViewHeader,
    IAccordionViewHeaderProps,
} from '@sberbusiness/triplex/components/AccordionView/components/AccordionViewHeader';
import {AccordionViewBody, IAccordionViewBodyProps} from '@sberbusiness/triplex/components/AccordionView/components/AccordionViewBody';
import {TestProps} from '@sberbusiness/triplex/types/CoreTypes';
import {IAccordionBodyProvideProps, IAccordionHeaderProvideProps} from '@sberbusiness/triplex/components/AccordionBase/types';
import {AccordionBase} from '../AccordionBase/protected/AccordionBase';

/** Свойства компонента AccordionView. */
interface AccordionViewPros extends React.HTMLAttributes<HTMLDivElement>, TestProps {
    children: [React.ReactElement<IAccordionViewHeaderProps>, React.ReactElement<IAccordionViewBodyProps>];
    /**
     * Флаг открытия/закрытия аккордеона, для внешнего управления.
     * Если пропс не передан, то управление открытием/закрытием происходит внутри компонента.
     */
    isOpen?: boolean;
    /**
     * Функция вызывающаяся после открытия/закрытия аккордеона пользователем.
     * isOpen - состояние открыт/закрыт после действий пользователя.
     */
    onToggle?: (isOpen: boolean) => void;
    /**
     * Функция вызывающаяся для открытия/закрытия аккордеона пользователем.
     * isOpen - состояние открыт/закрыт после действий пользователя.
     */
    toggle?: (isOpen: boolean) => void;
}

/** Компонент аккордеон, только для отображения пользователю информации. */
export class AccordionView extends React.Component<AccordionViewPros> {
    public static displayName = 'AccordionView';

    public static Header = AccordionViewHeader;
    public static Body = AccordionViewBody;

    public render(): JSX.Element {
        const {children, className, isOpen, onToggle, toggle, ...htmlAttributes} = this.props;

        return (
            <div {...htmlAttributes} className={classnames(className, 'cssClass[accordionView]')}>
                <AccordionBase
                    expandAnimationClassName="cssClass[accordionViewContent]"
                    renderBody={this.renderBody}
                    renderHeader={this.renderHeader}
                    isOpen={isOpen}
                    onToggle={onToggle}
                    toggle={toggle}
                />
            </div>
        );
    }

    private renderBody = (props: IAccordionBodyProvideProps) => {
        const [header, body] = this.props.children;
        return React.cloneElement(body, {...props});
    };

    private renderHeader = (props: IAccordionHeaderProvideProps) => {
        const [header] = this.props.children;
        return React.cloneElement(header, {...props});
    };
}
