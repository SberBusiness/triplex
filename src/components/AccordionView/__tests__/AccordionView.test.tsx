import {AccordionView} from '@sberbusiness/triplex/components/AccordionView/AccordionView';
import {AccordionBase} from '@sberbusiness/triplex/components/AccordionBase/protected/AccordionBase';
import {AccordionViewHeader} from '@sberbusiness/triplex/components/AccordionView/components/AccordionViewHeader';
import {allure} from '@jest/unit/allure-report';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import React from 'react';

jest.mock('@sberbusiness/icons/CaretdownSrvxIcon16', () => ({
    CaretdownSrvxIcon16: 'svg',
}));

jest.mock('@sberbusiness/triplex/utils/uniqueId', () => ({
    uniqueId: jest.fn(() => 0),
}));

describe('AccordionView', () => {
    beforeEach(() => {
        allure.feature('AccordionView');
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <AccordionView>
                    <AccordionView.Header>Test</AccordionView.Header>
                    <AccordionView.Body>Test</AccordionView.Body>
                </AccordionView>
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('calls onToggle', () => {
        const handleToggle = jest.fn();
        const wrapper = shallow(
            <AccordionView onToggle={handleToggle}>
                <AccordionView.Header>Test</AccordionView.Header>
                <AccordionView.Body>Test</AccordionView.Body>
            </AccordionView>
        );

        wrapper.find(AccordionBase).simulate('toggle');
        expect(handleToggle).toHaveBeenCalled();
    });

    it('opens correctly', () => {
        const component = renderer.create(
            <AccordionView>
                <AccordionView.Header>Test</AccordionView.Header>
                <AccordionView.Body>Test</AccordionView.Body>
            </AccordionView>
        );

        component.root.findByType(AccordionViewHeader).props.toggle();
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('closes correctly', () => {
        const component = renderer.create(
            <AccordionView>
                <AccordionView.Header>Test</AccordionView.Header>
                <AccordionView.Body>Test</AccordionView.Body>
            </AccordionView>
        );

        component.root.findByType(AccordionViewHeader).props.toggle();
        component.root.findByType(AccordionViewHeader).props.toggle();
        expect(component.toJSON()).toMatchSnapshot();
    });
});
