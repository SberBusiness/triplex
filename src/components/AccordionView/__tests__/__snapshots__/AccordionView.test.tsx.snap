// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionView closes correctly 1`] = `
<div
  className="cssClass[accordionView]"
>
  <div
    aria-controls="AccordionBase-0"
    aria-expanded="false"
    className="cssClass[accordionViewHeaderWrapper] cssClass[animating]"
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <h5
      className="cssClass[accordionViewHeader]"
      title="Test"
    >
      Test
    </h5>
    <svg
      className="cssClass[caretIcon]"
    />
  </div>
  <div
    className="cssClass[accordionViewContent] cssClass[globalExpandAnimationAnimating]"
    id="AccordionBase-0"
    onTransitionEnd={[Function]}
    style={
      {
        "display": "none",
      }
    }
  >
    <div>
      <div
        className="cssClass[accordionViewBody]"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`AccordionView opens correctly 1`] = `
<div
  className="cssClass[accordionView]"
>
  <div
    aria-controls="AccordionBase-0"
    aria-expanded="true"
    className="cssClass[accordionViewHeaderWrapper] cssClass[animating] cssClass[opened]"
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <h5
      className="cssClass[accordionViewHeader]"
      title="Test"
    >
      Test
    </h5>
    <svg
      className="cssClass[caretIcon]"
    />
  </div>
  <div
    className="cssClass[accordionViewContent] cssClass[globalExpandAnimationAnimating] cssClass[globalExpandAnimationOpened]"
    id="AccordionBase-0"
    onTransitionEnd={[Function]}
    style={
      {
        "display": "none",
      }
    }
  >
    <div>
      <div
        className="cssClass[accordionViewBody]"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`AccordionView renders correctly 1`] = `
<div
  className="cssClass[accordionView]"
>
  <div
    aria-controls="AccordionBase-0"
    aria-expanded="false"
    className="cssClass[accordionViewHeaderWrapper]"
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    role="button"
    tabIndex={0}
  >
    <h5
      className="cssClass[accordionViewHeader]"
      title="Test"
    >
      Test
    </h5>
    <svg
      className="cssClass[caretIcon]"
    />
  </div>
  <div
    className="cssClass[accordionViewContent]"
    id="AccordionBase-0"
    onTransitionEnd={[Function]}
    style={
      {
        "display": "none",
      }
    }
  >
    <div>
      <div
        className="cssClass[accordionViewBody]"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;
